crates/webidl/src/util.rs

```diff
@@ -707,3 +712,73 @@ pub fn public() -> syn::Visibility {
         pub_token: Default::default(),
     })
 }
+
+
+/// When generating our web_sys APIs we default to setting slice references that
+/// get passed to JS as mutable in case they get mutated in JS.
+///
+/// In certain cases we know for sure that the slice will not get mutated - for
+/// example when working with the WebGlRenderingContext APIs.
+///
+/// Here we implement a whitelist for those cases. This whitelist is currently
+/// maintained by hand.
+///
+/// When adding to this whitelist add tests to crates/web-sys/tests/wasm/whitelisted_immutable_slices.rs
+fn maybe_adjust<'a>(mut idl_type: IdlType<'a>, id: &'a OperationId) -> IdlType<'a> {
+    let op = match id {
+        OperationId::Operation(Some(op)) => op,
+        _ => return idl_type
+    };
+
+    if IMMUTABLE_F32_WHITELIST.contains(op) {
+        flag_slices_immutable(&mut idl_type)
+    }
+
+    // TODO: Add other whitelisted slices here, such as F64 or u8..
+
+    idl_type
+
+}
+
+fn flag_slices_immutable(ty: &mut IdlType) {
+    match ty {
+        IdlType::Float32Array { immutable } => *immutable = true,
+        IdlType::Union(list) => {
+            for item in list { flag_slices_immutable(item); }
+        }
+
+        // TODO: ... other recursive cases like Nullable handled here
```

Comment:
I think there's only a few of these other than `Union`, mind going ahead and expanding them in this PR?

Reply:
Ohhh gotcha - will do next time I hop in this branch!